### PR TITLE
feat(db-engine): Add support for Apache Solr

### PR DESCRIPTION
**Pull request by aadel**
### SUMMARY
Adding support for Apache Solr through SQLAlchemy DBAPI subject to features provided by and limitations of Solr Parallel SQL interface. An engine spec is included.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Example Charts:

Screen Shot 2021-01-10 at 6 31 27 AM
Screen Shot 2021-01-10 at 6 36 21 AM
Screen Shot 2021-01-10 at 6 44 20 AM
Screen Shot 2021-01-10 at 6 50 08 AM
Screen Shot 2021-01-10 at 7 06 54 AM
Screen Shot 2021-01-10 at 7 12 50 AM

### TEST PLAN
Most of the testing is maintained on `sqlalchemy-solr` package since the current engine spec is skeletal.

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by villebro — APPROVED** [review]
This is awesome @aadel, thanks so much for adding support for Solr! 🎉  LGTM, but requesting review from @srinify to check the docs changes.

**Comment by aadel** [comment]
Thanks @villebro! I'm also excited adding support for Solr! 💯

**Comment by amitmiran137** [comment]
Is there a plan on the roadmap to support facets?
How would a SQL query would look like for that ?

**Comment by aadel** [comment]
@amitmiran137 Actually the concept of facets is associated with Lucene/Solr however, the SQL interface uses supports aggregations, predicates and sorting using standard syntax which give the same results for supported scenarios. Internally, the SQL interface in the default mode uses the JSON Facet API or StatsComponent for aggregations. In this mode, the aggregations logic is pushed down into the search engine and the aggregates are returned.

**Comment by willbarrett** [comment]
Nice! Added @dpgaspar as a reviewer as well.

**Review by dpgaspar — APPROVED** [review]
Awesome!

File: superset/db_engine_specs/solr.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+# under the License.
+
+from superset.db_engine_specs.base import BaseEngineSpec
+
+
+class SolrEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
+    """Engine spec for Apache Solr"""
+
+    engine = "solr"
+    engine_name = "Apach Solr"
```
Comment (duynguyenhoang) [inline superset/db_engine_specs/solr.py:25]: nit: typo here. It should be "Apache"

Comment (villebro) [reply superset/db_engine_specs/solr.py:25]: Thanks for catching @duynguyenhoang ! This would be good to fix before merging.

**Review by villebro — CHANGES_REQUESTED** [review]
Request to fix typo that was picked up by @duynguyenhoang

File: superset/db_engine_specs/solr.py (hunk 1 above)
Comment (srinify) [reply superset/db_engine_specs/solr.py:25]: hey @aadel this PR looks good minus this one typo! "Apach" -> "Apache"

Fix this and I'd be happy to approve!

**Review by srinify — CHANGES_REQUESTED** [review]

**Review by villebro — APPROVED** [review]

**Pull request merged**